tutorial.md

```diff
@@ -0,0 +1,225 @@
+makes it possible to deal with expressions that would otherwise be too large to
+write by hand, and opens possibilities to perform automatic mathematical
+optimizations of expressions.
+
+SymPy supports generating code for C, C++, Fortran, Matlab/Octave, Python,
+Cython, Julia, Javascript, LLVM, Rust, Haskell, Mathematica, Tensorflow, and
+Theano, and can easily be extended to other languages. SymPy’s code generation
+is used by libraries such as PyDy, pyodesys, sympybotics, pycalphad, and many
+other programs.
+
+## Learning objectives
+
+Attendees will be able to:
+- write SymPy expressions describing mathematical functions and identify the
+  function arguments and outputs.
+- use the SymPy code printers to transform SymPy expressions representing
+  common domain specific functions into multiple output languages.
+- use the SymPy code generation routines to output compilable C code and use
+  Cython to access these functions in Python.
+- generate custom vectorized functions with the three SymPy functions:
+  lambdify, ufuncify, and autowrap.
+- create both custom code printers that make use of specialized C libraries and
+  common subexpression elimination (CSE).
+- subclass the core SymPy printers and create a printer for a custom language.
+
+## Outline
+
+Intro to SymPy Expressions [30 minutes]
+
+- Description: Writing common domain specific mathematical expressions with
+  SymPy.
+- Motivating Examples: Long expressions, Matrix operations, and Loop Fusion
+  from classical mechanics, chemical kinetics, nuclear dynamics, and materials
+  science.
+
+Code Printers [30 minutes]
+
+- Description: Printing expressions in multiple languages (C, Fortran, Rust,
+  Julia, Octave, Javascript, etc)
+- Motivating Example: 2D interactive plot in a Jupyter notebook by javascript
+  injection
+
+The Easy Way: High Level Generation (lambdify, ufuncify) [30 minutes]
+
+- Description: Generate loop fused NumPy ufuncs and compare to automatically
+  generated NumPy code. Show how you can extend lambdify with custom Python
+  functions.
+- Motivating Example: Generate a Jacobian function for a chemical kinetic
+  problem.
+
+The Harder Way: Code generation and compilation [1 hour]
+
+- Description: Write a tight low level loop with an indexed type and a long
+  expressions with knowing C.
+- Motivating Example: Evaluate the chemical kinetic Jacobian in a loop.
+
+Cythonizing Your Code (manually and autowrap) [30 minutes]
+
+- Description: Generate C code to evaluate the gradient and Jacobian of an
+  ordinary differential equation and wrap it for use with SciPy’s odeint
+  function.
+- Motivating Example: Chaotic triple pendulum example from classical mechanics.
+
+Extending SymPy’s classes [1 hour]
+
+- Description: Show how to use external C libraries and optimize your code with
+  common sub-expression elimination.
+- Motivating Example: Speed up triple pendulum execution with CSE and GSL
```

Comment:
Taking the code from "The harder way" we can adjust so that it provides a sundials compatible interface. That way the function body is reused, we only add common sub-expression elimination and accessor macros to the LHS.